make k8sClient more resilient to slow k8s-api response (httpConnection config)

naineshmp opened the pull request:
- add config to update httpConnectionReadTimeout (also set default)
- add config to update httpConnectionTimeout (also set default)

naineshmp commented:
we have noticed in our project that under stress k8s api responses are taking more than default httpConnectionTimeouts (1-2 seconds) and workflows fail with `SocketTimeoutException`. 

After updating readTimeout (timeout to receive first byte) and connectionTimeout (timeout to establish connection) - k8sClient is more resilient to transient k8s-api slowness. Hoping this might be helpful to other users as well. I have also added config and doc, let me know if any additional changes required or if there is better way. 

@pditommaso 

```groovy
        conn.setReadTimeout(config.requestHttpReadTimeoutInMilliSeconds)
        conn.setConnectTimeout(config.requestHttpConnectionTimeoutInMilliSeconds)
```

pditommaso commented on docs/config.rst:
```diff
@@ -423,6 +423,8 @@ storageSubPath      The path in the persistent volume to be mounted (default: ro
 computeResourceType Define whether use Kubernetes ``Pod`` or ``Job`` resource type to carry out Nextflow tasks (default: ``Pod``).
 fetchNodeName       If you trace the hostname, activate this option (default: ``false``, requires version ``22.05.0-edge`` or later).
 volumeClaims        (deprecated)
+requestHttpReadTimeoutInMilliSeconds    Defines the Kubernetes client request http connection read timeout (default 60000ms)
```
```suggestion
httpReadTimeout    Defines the Kubernetes client request HTTP connection read timeout e.g. ``'60s'``
```

pditommaso commented on docs/config.rst:
```diff
@@ -423,6 +423,8 @@ storageSubPath      The path in the persistent volume to be mounted (default: ro
 fetchNodeName       If you trace the hostname, activate this option (default: ``false``, requires version ``22.05.0-edge`` or later).
 volumeClaims        (deprecated)
+requestHttpReadTimeoutInMilliSeconds    Defines the Kubernetes client request http connection read timeout (default 60000ms)
+requestHttpConnectionTimeoutInMilliSeconds  Defines the Kubernetes client request http connection timeout (default 60000ms)
```
```suggestion
httpConnectionTimeout  Defines the Kubernetes client request HTTP connection timeout e.g. ``'60s'``'
```

Nextflow usually uses [Duration](https://github.com/nextflow-io/nextflow/blob/84d5a747d1ac4baa425313f8e16e0c3efa6a642b/modules/nf-commons/src/main/nextflow/util/Duration.groovy#L35-L35) class for those setting

pditommaso commented on modules/nextflow/src/main/groovy/nextflow/k8s/client/ClientConfig.groovy:
```diff
@@ -53,6 +53,14 @@ class ClientConfig {
 
+    //Timeout when reading from Input stream when a connection is established to a resource.
+    //If the timeout expires before there is data available for read, a java.net. SocketTimeoutException is raised
+    Integer requestHttpReadTimeoutInMilliSeconds = 60 * 1000 //60 seconds
```
```suggestion
    Duration httpReadTimeout
```

Default should be null for backward compatibility

pditommaso commented on modules/nextflow/src/main/groovy/nextflow/k8s/client/ClientConfig.groovy:
```diff
@@ -53,6 +53,14 @@ class ClientConfig {
+
+    //Timeout when opening a communications link to the resource referenced by K8sClient request connection
+    //If the timeout expires before there is data available for read, a java.net. SocketTimeoutException is raised
+    Integer requestHttpConnectionTimeoutInMilliSeconds = 60 * 1000 //60 seconds
```
```suggestion
    Duration httpConnectionTimeout 
```

pditommaso reviewed (changes requested):
Thanks for submitting this improvement. I've made a few comments. Also a unit test should be added to verify those setting are correcting applied 

You can following this snippet as reference 

https://github.com/nextflow-io/nextflow/blob/84d5a747d1ac4baa425313f8e16e0c3efa6a642b/modules/nextflow/src/test/groovy/nextflow/k8s/K8sConfigTest.groovy#L131-L144

naineshmp commented:
> Thanks for submitting this improvement. I've made a few comments. Also a unit test should be added to verify those setting are correcting applied
> 
> You can following this snippet as reference
> 
> https://github.com/nextflow-io/nextflow/blob/84d5a747d1ac4baa425313f8e16e0c3efa6a642b/modules/nextflow/src/test/groovy/nextflow/k8s/K8sConfigTest.groovy#L131-L144

Thank you! I have just added the changes and added positive and negative tests. Please let me know if any more changes required.

pditommaso commented on modules/nextflow/src/test/groovy/nextflow/k8s/K8sConfigTest.groovy:
```diff
@@ -140,6 +141,29 @@ class K8sConfigTest extends Specification {
+        def CONFIG = [
+                namespace: 'this',
+                serviceAccount: 'that',
+                client: [server: 'http://foo', httpReadTimeout: '20s', httpConnectionTimeout: '25s']
```
I think it would be better to keep those settings in the top-most scope (like `namespace`) otherwise they will be ignored when the client is built via the (kube) context. See here 

https://github.com/nextflow-io/nextflow/blob/master/modules/nextflow/src/main/groovy/nextflow/k8s/K8sConfig.groovy#L210-L221

pditommaso reviewed (changes requested):
Thanks for improving this PR. I've made another comment. Please consider to sign the commit as specified here https://github.com/nextflow-io/nextflow/blob/master/CONTRIBUTING.md#contributing-code-changes

naineshmp commented:
> Thanks for improving this PR. I've made another comment. Please consider to sign the commit as specified here https://github.com/nextflow-io/nextflow/blob/master/CONTRIBUTING.md#contributing-code-changes

Thanks for the update. I have made the changes. Do I need to sign all the previous commits as well?

pditommaso commented:
Thanks a lot for this contribution. I've changed `httpConnectionTimeout` to `httpConnectTimeout` and merged it manually. See 064f9bc4.
